gui/src/WorkspaceView/controls/SurfaceSelectionControls.tsx

```diff
@@ -0,0 +1,164 @@
+    selectedVectorFieldCallback: any
+    selectedScalarFieldCallback: any
+}
+
+type FieldSelectorProps = {
+    surfaceName: string
+    fields: WorkspaceSurfaceScalarField[] | WorkspaceSurfaceVectorField[]
+    selectedFieldName?: string
+    type: 'vector' | 'scalar'
+    callback: any
+}
+
+type ChangeEvent = React.ChangeEvent<{value: unknown}>
+
+const SurfaceSelectionControls: FunctionComponent<SurfaceSelectionControlProps> = (props: SurfaceSelectionControlProps) => {
+    const { data, selection, selectionDispatch } = props
+
+    const handleToggleVisibleSurface = useCallback((e: ChangeEvent) => {
+        const surfaceName = e.target.value as string
+        selectionDispatch({
+            type: 'toggleVisibleSurface',
+            surfaceName
+        })
+    }, [selectionDispatch])
+
+    const fieldTtoggler = useCallback((name: string, fieldType: 'scalar' | 'vector') => {
+        const type = fieldType === 'scalar' ? 'toggleSelectedSurfaceScalarField' : 'toggleSelectedSurfaceVectorField'
+        const [ surfaceName, surfaceFieldName ] = name.split(separator)
+        selectionDispatch({ type, surfaceName, surfaceFieldName })
+    }, [selectionDispatch])
+
+    const handleToggleSelectedScalarField = useCallback((e: ChangeEvent) => {
+        fieldTtoggler(e.target.value as string, 'scalar')
+    }, [fieldTtoggler])
+
+    const handleToggleSelectedVectorField = useCallback((e: ChangeEvent) => {
+        fieldTtoggler(e.target.value as string, 'vector')
+    }, [fieldTtoggler])
+
+    // This is probably not a huge performance boost, but we expect the underlying data to change rarely,
+    // while the selection state will change much more frequently.
+    // Might as well memoize the filtering that matches surfaces to fields, so we don't need to re-filter
+    // every time a selection state changes.
+    const surfacesWithFields = useMemo(() => (
+        data.surfaces.map(surface => {
+            return {
+                name: surface.name,
+                scalarFields: data.surfaceScalarFields.filter(field => field.surfaceName === surface.name),
+                vectorFields: data.surfaceVectorFields.filter(field => field.surfaceName === surface.name)
+            }
+        })
+    ), [data.surfaces, data.surfaceScalarFields, data.surfaceVectorFields])
+
+    const surfaces = useMemo(() => (surfacesWithFields.map(surface => {
+        return {
+            name: surface.name,
+            displayed: (selection.visibleSurfaceNames || []).includes(surface.name),
+            scalarFields: surface.scalarFields,
+            vectorFields: surface.vectorFields,
+            selectedScalarFieldName: selection.selectedSurfaceScalarFieldNames[surface.name],
+            selectedVectorFieldName: selection.selectedSurfaceVectorFieldNames[surface.name]
+        }
+    })), [surfacesWithFields, selection.selectedSurfaceVectorFieldNames, selection.selectedSurfaceScalarFieldNames, selection.visibleSurfaceNames])
+
```

Comment:
As the comment l. 61-64 says, I split the memoization into two parts. I doubt the data sets will be large enough that this will ever make a difference, but it just rubbed me the wrong way.

Anyway, the first one looks only at the `data` parameter and matches fields to their surfaces, while the second looks at selection state.